### PR TITLE
[SPARK-29586][SQL]changge jdbc method param lowerBound and upperBound DataType

### PR DESCRIPTION
**Pull request by Le-Dai**
### What changes were proposed in this pull request?

changge param dataType

### Why are the changes needed?

 partionColum timestamp and date dataType is not support

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Thank you for making a PR, @MrDLontheway .
Could you update the PR description with the reason why we need this change?

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala [hunk 1]
```diff
@@ -491,7 +491,14 @@ class JDBCSuite extends QueryTest
...
   test("Partitioning via JDBCPartitioningInfo API") {
-    val df = spark.read.jdbc(urlWithUserAndPass, "TEST.PEOPLE", "THEID", 0, 4, 3, new Properties())
+    val df = spark.read.jdbc(urlWithUserAndPass,
+      "TEST.PEOPLE",
+      "THEID",
+      "0",
+      "4",
+      3,
+      new Properties()
+    )
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: Let's change like the following.
```scala
val df =
  spark.read.jdbc(urlWithUserAndPass, "TEST.PEOPLE", "THEID", 0, 4, 3, new Properties())
```

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Also, we need a new test case for the new features. Currently, this PR only adjusted old test cases.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala [hunk 2]
```diff
@@ -284,15 +284,15 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
       url: String,
       table: String,
       columnName: String,
-      lowerBound: Long,
-      upperBound: Long,
+      lowerBound: String,
+      upperBound: String,
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala]: BTW, this suggestion is a breaking change. We need to accept both one if we want to proceed.

cc @cloud-fan and @gatorsmile

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala [hunk 3]
```diff
@@ -1658,4 +1673,56 @@ class JDBCSuite extends QueryTest
...
+
+    df2.logicalPlan match {
+      case LogicalRelation(JDBCRelation(_, parts, _), _, _, _) =>
+        val whereClauses = parts.map(_.asInstanceOf[JDBCPartition].whereClause).toSet
+        assert(whereClauses === Set(
+          """"T" < '2018-07-15 20:50:32.5' or "T" is null""",
+          """"T" >= '2018-07-15 20:50:32.5'"""))
+    }
+    checkAnswer(df2, expectedResult)
+  }
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: nit: It seems the most part of this test is the same with https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala#L1517-L1565
So, can we share it by using a helper func?

**Comment by maropu** [comment]
The support itself looks nice to me.

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: Ping, @MrDLontheway , for the above comment.

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala [hunk 4]
```diff
@@ -504,17 +511,25 @@ class JDBCSuite extends QueryTest
...
-    val df2 = spark.read.jdbc(urlWithUserAndPass, "TEST.EMP", "THEID", 0, 4, 3, new Properties())
+    val df2 = spark.read.jdbc(
+      urlWithUserAndPass,
+      "TEST.EMP",
+      "THEID",
+      "0",
+      "4",
+      3,
+      new Properties()
+    )
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: ditto.
```scala
val df2 =
  spark.read.jdbc(urlWithUserAndPass, "TEST.EMP", "THEID", "0", "4", 3, new Properties())
```

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala (hunk 3 above)
Comment (MrDLontheway) [reply sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: I reused it to verify the correctness of the method.

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala (hunk 1 above)
Comment (MrDLontheway) [reply sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: thanks for reminding

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala (hunk 3 above)
Comment (MrDLontheway) [reply sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: @maropu 
I didn't think of a good way, maybe maybe I can combine them like this
```
test("SPARK-22814 support date/timestamp types in partitionColumn") {
    val expectedResult = Seq(
      ("2018-07-06", "2018-07-06 05:50:00.0"),
      ("2018-07-06", "2018-07-06 08:10:08.0"),
      ("2018-07-08", "2018-07-08 13:32:01.0"),
      ("2018-07-12", "2018-07-12 09:51:15.0")
    ).map { case (date, timestamp) =>
      Row(Date.valueOf(date), Timestamp.valueOf(timestamp))
    }

    // DateType partition column
    val df1 = spark.read.format("jdbc")
      .option("url", urlWithUserAndPass)
      .option("dbtable", "TEST.DATETIME")
      .option("partitionColumn", "d")
      .option("lowerBound", "2018-07-06")
      .option("upperBound", "2018-07-20")
      .option("numPartitions", 3)
      .load()

    val df2 = spark.read.jdbc(
      url = urlWithUserAndPass,
      table = "TEST.DATETIME",
      columnName = "d",
      lowerBound = "2018-07-06",
      upperBound = "2018-07-20",
      numPartitions = 3,
      connectionProperties = new Properties()
    )

    dateMatch(df1)
    dateMatch(df2)
    checkAnswer(df1, expectedResult)
    checkAnswer(df2, expectedResult)

    def dateMatch(df: DataFrame): Unit ={
      df.logicalPlan match {
        case LogicalRelation(JDBCRelation(_, parts, _), _, _, _) =>
          val whereClauses = parts.map(_.asInstanceOf[JDBCPartition].whereClause).toSet
          assert(whereClauses === Set(
            """"D" < '2018-07-10' or "D" is null""",
            """"D" >= '2018-07-10' AND "D" < '2018-07-14'""",
            """"D" >= '2018-07-14'"""))
      }
    }

    // TimestampType partition column
    val df3 = spark.read.format("jdbc")
      .option("url", urlWithUserAndPass)
      .option("dbtable", "TEST.DATETIME")
      .option("partitionColumn", "t")
      .option("lowerBound", "2018-07-04 03:30:00.0")
      .option("upperBound", "2018-07-27 14:11:05.0")
      .option("numPartitions", 2)
      .load()

    val df4 = spark.read.jdbc(
      url = urlWithUserAndPass,
      table = "TEST.DATETIME",
      columnName = "t",
      lowerBound = "2018-07-04 03:30:00.0",
      upperBound = "2018-07-27 14:11:05.0",
      numPartitions = 2,
      connectionProperties = new Properties()
    )

    timeStampMatch(df3)
    timeStampMatch(df4)
    checkAnswer(df3, expectedResult)
    checkAnswer(df4, expectedResult)

    def timeStampMatch(df: DataFrame): Unit ={
      df.logicalPlan match {
        case LogicalRelation(JDBCRelation(_, parts, _), _, _, _) =>
          val whereClauses = parts.map(_.asInstanceOf[JDBCPartition].whereClause).toSet
          assert(whereClauses === Set(
            """"T" < '2018-07-15 20:50:32.5' or "T" is null""",
            """"T" >= '2018-07-15 20:50:32.5'"""))
      }
    }
  }
```

**Comment by cloud-fan** [comment]
We can't change the public API. I think the current API is OK as a shortcut to specify upper/lower bound for integral type column (we can improve the doc). User can specify whatever bound values they want via `DataFrameReader.option`, right?

**Pull request closed without merging**